DigitalLearningSolutions.Data/Services/GroupsService.cs

```diff
@@ -176,23 +176,24 @@ CentreAnswersData newCentreAnswers
 
             foreach (var changedAnswer in changedLinkedFields)
             {
-                var groupToRemoveDelegateFrom = allSynchronisedGroupsAtCentre.SingleOrDefault(
+                var groupsToRemoveDelegateFrom = allSynchronisedGroupsAtCentre.Where(
                     g => g.LinkedToField == changedAnswer.LinkedFieldNumber &&
                          GroupLabelMatchesAnswer(g.GroupLabel, changedAnswer.OldValue, changedAnswer.LinkedFieldName)
-                );
+                ).ToList();
 
-                var groupToAddDelegateTo = allSynchronisedGroupsAtCentre.SingleOrDefault(
+                var groupsToAddDelegateTo = allSynchronisedGroupsAtCentre.Where(
                     g => g.LinkedToField == changedAnswer.LinkedFieldNumber &&
                          GroupLabelMatchesAnswer(g.GroupLabel, changedAnswer.NewValue, changedAnswer.LinkedFieldName)
-                );
+                ).ToList();
 
                 using var transaction = new TransactionScope();
-                if (groupToRemoveDelegateFrom != null)
+
+                foreach (var groupToRemoveDelegateFrom in groupsToRemoveDelegateFrom)
                 {
                     RemoveDelegateFromGroup(delegateAccountWithOldDetails.Id, groupToRemoveDelegateFrom.GroupId);
                 }
 
-                if (groupToAddDelegateTo != null)
+                foreach (var groupToAddDelegateTo in groupsToAddDelegateTo)
                 {
                     groupsDataService.AddDelegateToGroup(
                         delegateAccountWithOldDetails.Id,

```

Comment:
The other bug was that SynchroniseUserChangesWithGroups only expected add/remove the user from one group, but the group generation process lets you create many groups from the same registration field, so I changed this to account for that. Shout if you think this should be in a separate MR

Reply:
Eh, we really want to have some extra tests for this functionality. So there's a bit more work to do, so yeah, I think putting it in a new PR is sensible. I've assigned you the ticket, if you could move this change over to another branch?

Reply:
https://softwiretech.atlassian.net/browse/HEEDLS-805
Just realised you might not have known - Kevin spotted some iffy logs and asked me to investigate, resulting in this ticket. Well done for spotting it yourself!

Reply:
Reverted the last commit with those changes and moved them to a new MR https://github.com/TechnologyEnhancedLearning/DLSV2/pull/981